util/winutil/svcdiag_windows.go

```diff
@@ -0,0 +1,297 @@
+// Copyright (c) 2022 Tailscale Inc & AUTHORS All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package winutil
+
+import (
+	"encoding/hex"
+	"encoding/json"
+	"fmt"
+	"strings"
+	"unsafe"
+
+	"golang.org/x/sys/windows"
+	"golang.org/x/sys/windows/svc"
+	"golang.org/x/sys/windows/svc/mgr"
+	"tailscale.com/types/logger"
+)
+
+// LogSvcState obtains the state of the Windows service named rootSvcName and
+// all of its dependencies, and then emits that state to logf.
+func LogSvcState(logf logger.Logf, rootSvcName string) {
+	logEntries := []svcStateLogEntry{}
+
+	walkFn := func(svc *mgr.Service, config mgr.Config) {
+		status, err := svc.Query()
+		if err != nil {
+			logf("Failed retrieving Status for service %q: %v", svc.Name, err)
+		}
+
+		logEntries = append(logEntries, makeLogEntry(svc, status, config))
+	}
+
+	err := walkServices(rootSvcName, walkFn)
+	if err != nil {
+		logf("LogSvcState: %v", err)
+		return
+	}
+
+	json, err := json.MarshalIndent(logEntries, "", "  ")
+	if err != nil {
+		logf("Error marshaling service log entries: %v", err)
+		return
+	}
+
+	var builder strings.Builder
+	builder.WriteString(`State of service "`)
+	builder.WriteString(rootSvcName)
+	builder.WriteString(`" and its dependencies:`)
+	builder.WriteString("\n")
+	builder.Write(json)
+	builder.WriteString("\n")
+
+	logf(builder.String())
+}
+
+// walkSvcFunc is type of the callback function invoked by WalkServices.
+type walkSvcFunc func(*mgr.Service, mgr.Config)
+
+// walkServices opens the service named rootSvcName and walks its dependency
+// graph, invoking callback for each service (including the root itself).
+func walkServices(rootSvcName string, callback walkSvcFunc) error {
+	scm, err := connectToLocalSCMForRead()
+	if err != nil {
+		return fmt.Errorf("failed connecting to Service Control Manager: %w", err)
+	}
+	defer scm.Disconnect()
+
+	rootSvc, err := openServiceForRead(scm, rootSvcName)
+	if err != nil {
+		return fmt.Errorf("failed opening service %q: %w", rootSvcName, err)
+	}
+
+	deps := []*mgr.Service{rootSvc}
+	seen := make(map[string]struct{})
+
+	for err == nil && len(deps) > 0 {
+		err = func() error {
+			curSvc := deps[0]
+			defer curSvc.Close()
+
+			deps = deps[1:]
+
+			seen[curSvc.Name] = struct{}{}
+
+			curCfg, err := curSvc.Config()
+			if err != nil {
+				return fmt.Errorf("failed retrieving Config for service %q: %w", curSvc.Name, err)
+			}
+
+			callback(curSvc, curCfg)
+
+			for _, depName := range curCfg.Dependencies {
+				if _, ok := seen[depName]; ok {
+					continue
+				}
+
+				depSvc, err := openServiceForRead(scm, depName)
+				if err != nil {
+					return fmt.Errorf("failed opening service %q: %w", depName, err)
+				}
+
+				deps = append(deps, depSvc)
```

Comment:
I think you want to set `seen[depSvc.Name]` here otherwise you can end up with multiple copies of the same not-yet-visited dependency?

Reply:
`depSvc.Name` == `depName` which we've already checked.